generics/arraypool/pool.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package arraypool
+
+import (
+	"github.com/m3db/m3x/pool"
+)
+
+// elemArrayPool provides a pool for elemType slices.
+type elemArrayPool interface {
+	Init()
+	Get() []elemType
+	Put(elems []elemType)
+}
+
+type elemFinalizeFn func([]elemType) []elemType
+
+type elemArrayPoolOpts struct {
+	Options     pool.ObjectPoolOptions
+	Capacity    int
+	MaxCapacity int
+	FinalizeFn  elemFinalizeFn
+}
+
+type elemArrPool struct {
+	opts elemArrayPoolOpts
+	pool pool.ObjectPool
+}
+
+// nolint
+func newElemArrayPool(opts elemArrayPoolOpts) elemArrayPool {
+	if opts.FinalizeFn == nil {
+		opts.FinalizeFn = defaultElemFinalizerFn
+	}
+	p := pool.NewObjectPool(opts.Options)
+	return &elemArrPool{opts, p}
+}
+
+func (p *elemArrPool) Init() {
+	p.pool.Init(func() interface{} {
+		return make([]elemType, 0, p.opts.Capacity)
+	})
+}
+
+func (p *elemArrPool) Get() []elemType {
+	return p.pool.Get().([]elemType)
+}
+
+func (p *elemArrPool) Put(arr []elemType) {
+	arr = p.opts.FinalizeFn(arr)
+	if max := p.opts.MaxCapacity; max > 0 && cap(arr) > max {
+		return
+	}
+	p.pool.Put(arr)
+}
+
+func defaultElemFinalizerFn(elems []elemType) []elemType {
+	var empty elemType
+	for i := range elems {
+		elems[i] = empty
+	}
+	elems = elems[:0]
+	return elems
+}
+
+// nolint
+type elemArr []elemType
```

Comment:
Is `elemArr` and `grow` used anywhere? Can't seem to find where they are used other than in the tests. If so, I'd suggest moving them to the test files to be clear of the scope.

Reply:
Yep not used in the current diff directly, included here because it's related to the slice type. E.g. https://github.com/m3db/m3db/blob/64f3241fed93fffecf484c73f43a8b478982bd9e/network/server/tchannelthrift/node/service.go#L1069 

It's better to have a single place for all the testing in case we need to make changes to the behaviour down the road. I'd prefer to leave here unless you feel strongly.

Reply:
I see, makes sense.

Do you envision the pool/array type will only be used as internal types/methods? It means the generated types will have to be placed in the same package where they are used and not outside the package. Not sure if that's too strong a restriction in terms of repo layout etc.

Reply:
I think it'll be internal only for majority of the cases but people will occasionally want to have an exported version. We support that through renaming (e.g. [steps](https://github.com/m3db/m3x/blob/fab025948308e135c2d03d1369e0ccca2e2f6e1e/Makefile#L172-L173)). 

TBH: I haven't tested creating an exported version yet, so the first person who does it will have to a make change or two for more renaming I'm sure.